### PR TITLE
REVSDL-2386 EventEngine integration

### PR DESCRIPTION
**Pull request by KKarlash**
@bonewell 
@VladSemenyuk 
@giangtqh 
@thinhts

### CODE REVIEW

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+template<typename EventMessage, typename EventID>
+Event<EventMessage, EventID>::Event(const EventMessage& message, const EventID& id)
+: event_message_(message),
+  id_(id) {
+}
+
+}  // namespace event_engine
+
+#endif /* SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_EVENT_ENGINE_EVENT_H_ */
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h]: // wrong comment style

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+   * @brief Default constructor
+   */
+  EventDispatcher();
+
+  /**
+   * @brief Destructor
+   */
+  virtual ~EventDispatcher();
+
+
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h]: empty line

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  typedef std::map<int32_t, ObserverList>      ObserversMap;
+  typedef std::map<EventID, ObserversMap>      EventObserverMap;
+
+  // Members section
+  sync_primitives::Lock                               state_lock_;
+  EventObserverMap                                    observers_;
+  DISALLOW_COPY_AND_ASSIGN(EventDispatcher);
+};
+
+
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h]: empty line

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+          ++observer_it;
+        }
+      }
+    }
+  }
+}
+
+}  // namespace event_engine
+
+#endif /* SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_EVENT_ENGINE_EVENT_DISPATCHER_H_ */
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h]: //

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h (hunk 1 above)
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h]: Incorrect type of comment. Should be `//`
You can use https://adc.luxoft.com/stash/projects/FORD/repos/toolkit/browse/cpplint/cpplint.py to check Google style

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+#include <map>
+
+#include "utils/lock.h"
+#include "utils/singleton.h"
+
+#include "can_cooperation/event_engine/event.h"
+#include "can_cooperation/event_engine/event_observer.h"
+
+#include "interfaces/HMI_API.h"
+
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h]: > Within each section the includes should be ordered alphabetically.
https://google.github.io/styleguide/cppguide.html#Names_and_Order_of_Includes

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+#define SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_EVENT_ENGINE_EVENT_DISPATCHER_H_
+
+#include <list>
+#include <map>
+
+#include "utils/lock.h"
+#include "utils/singleton.h"
+
+#include "can_cooperation/event_engine/event.h"
+#include "can_cooperation/event_engine/event_observer.h"
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h]: I believe you mean vr_cooperation :)

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h (hunk 4 above)
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h]: Incorrect type of comment. Should be `//`

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event_observer.h [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+template<typename EventMessage, typename EventID>
+EventObserver<EventMessage, EventID>::~EventObserver() {
+  // unsubscribe_from_all_events();
+}
+
+}  // namespace event_engine
+
+#endif /* SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_EVENT_ENGINE_EVENT_OBSERVER_H_ */
```
Comment (bonewell) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event_observer.h]: Incorrect type of comment. Should be //

**Comment by bonewell** [comment]
Reviewed.

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h (hunk 5 above)
Comment (KKarlash) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event_dispatcher.h]: @bonewell could you please clarify which section is not appropriate to this rule?

**Comment by giangtqh** [comment]
reviewed

**Comment by VladSemenyuk** [comment]
Reviewed

**Comment by thinhts** [comment]
reviewed

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  return id_;
+}
+
+template<typename EventMessage, typename EventID>
+const EventMessage& Event<EventMessage, EventID>::event_message() const {
+  return event_message_;
+}
+
+template<typename EventMessage, typename EventID>
+Event<EventMessage, EventID>::Event(const EventMessage& message, const EventID& id)
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h]: did you check with cpplint, line should be <= 80

File: src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+: event_message_(message),
+  id_(id) {
+}
+
+}  // namespace event_engine
+
+#endif // SRC_COMPONENTS_VR_COOPERATION_INCLUDE_VR_COOPERATION_EVENT_ENGINE_EVENT_H_
```
Comment (giangtqh) [inline src/components/vr_cooperation/include/vr_cooperation/event_engine/event.h]: Two spaces between the code and the comment.

**Pull request merged**